hm_fashion_recs/fe.py

```diff
@@ -0,0 +1,247 @@
+            transactions.query("@week <= week < @week_end")
+            .groupby("user")
+            .size()
+            .reset_index(name="user_volume")
+        )
+        df = df.merge(tmp, on="user", how="left")
+
+    with tm.timeit("user-item freshness features"):
+        tmp = (
+            transactions.query("@week <= week")
+            .groupby(["user", "item"])["day"]
+            .min()
+            .reset_index(name="user_item_day_min")
+        )
+        tmp["user_item_day_min"] -= transactions.query("@week == week")["day"].min()
+        df = df.merge(tmp, on=["item", "user"], how="left")
+
+    with tm.timeit("user-item volume features"):
+        week_end = week + CFG.user_item_volume_feature_weeks
+        tmp = (
+            transactions.query("@week <= week < @week_end")
+            .groupby(["user", "item"])
+            .size()
+            .reset_index(name="user_item_volume")
+        )
+        df = df.merge(tmp, on=["user", "item"], how="left")
+
+    with tm.timeit("item age volume features"):
+        week_end = week + CFG.age_volume_feature_weeks
+        tr = transactions.query("@week <= week < @week_end")[["user", "item"]].merge(
+            users[["user", "age"]], on="user"
+        )
+        item_age_volumes = []
+        for age in range(16, 100):
+            low = age - age_shifts[age]
+            high = age + age_shifts[age]
+            tmp = (
+                tr.query("@low <= age <= @high")
+                .groupby("item")
+                .size()
+                .reset_index(name="age_volume")
+            )
+            tmp["age_volume"] = tmp["age_volume"].rank(ascending=False)
+            tmp["age"] = age
+            item_age_volumes.append(tmp)
+        item_age_volumes = pd.concat(item_age_volumes)
+        df = df.merge(item_age_volumes, on=["item", "age"], how="left")
+
+    with tm.timeit("user category most frequent"):
+        for c in ["department_no_idx"]:
+            tmp = pd.read_pickle(user_features_path / f"user_ohe_agg_week{week}_{c}.pkl")
+            cols = [c for c in tmp.columns if c != "user"]
+            # tmp = tmp[['user'] + cols]
+            tmp[cols] = tmp[cols] / tmp[cols].mean()
+            tmp[f"{c}_most_freq_idx"] = np.argmax(tmp[cols].values, axis=1)
+            df = df.merge(tmp[["user", f"{c}_most_freq_idx"]])
+
+    with tm.timeit("ohe dot products"):
+        item_target_cols = [c for c in items.columns if c.endswith("_idx")]
+
+        items_with_ohe = pd.get_dummies(
+            items[["item"] + item_target_cols], columns=item_target_cols
+        )
+
+        cols = [c for c in items_with_ohe.columns if c != "item"]
+        items_with_ohe[cols] = items_with_ohe[cols] / items_with_ohe[cols].mean()
+
+        users_with_ohe = users[["user"]]
+        for c in item_target_cols:
+            tmp = pd.read_pickle(user_features_path / f"user_ohe_agg_week{week}_{c}.pkl")
+            assert tmp["user"].tolist() == users_with_ohe["user"].tolist()
```

Comment:
hope nobody minds if we rewrite it like this :)
```suggestion
            assert tmp["user"].equals(users_with_ohe["user"])
```

Reply:
I think assert is fine to rewrite